File: src/models/Pixel.ts
import Country from "./Country";

class BasePixel {
    x: number;
    y: number;
    
    constructor(x: number, y: number) {
        this.x = x;
        this.y = y;
    }
}

class Pixel extends BasePixel {
    country: Country | null;
    adjacentPixels: Pixel[];
    cooldownCap: number;
    cooldown: number;
    canBuildBoat: boolean;
    
    constructor(x: number, y: number, country: Country | null = null, adjacentPixels: Pixel[] = [], cooldown: number = 0, canBuildBoat: boolean = true) {
        super(x, y);
        this.country = country;
        this.adjacentPixels = adjacentPixels;
        this.cooldownCap = cooldown;
        this.cooldown = cooldown
        this.canBuildBoat = canBuildBoat;
    }

    updateWith(pixel: Pixel) {
        this.x = pixel.x;
        this.y = pixel.y;
        this.country = pixel.country;
        this.cooldownCap = pixel.cooldownCap;
        this.cooldown = pixel.cooldown;
        this.canBuildBoat = pixel.canBuildBoat;
    }

    addAdjacentPixel(pixel: Pixel) {
        this.adjacentPixels.push(pixel);
    }

    removeAdjacentPixel(pixel: Pixel) {
        try {
            const index = this.adjacentPixels.indexOf(pixel);
            if (index > -1)
                this.adjacentPixels.splice(index, 1);
        }
        catch {
            console.log("Adjacent pixel not found!");
        }
    }

    static copyPixel(pixel) {
        return new Pixel(pixel.x, pixel.y, pixel.country, [], pixel.cooldown, pixel.canBuildBoat);
    }
}

export default Pixel;

File: _python/models/pixel.py
# import copy
from models.country import Country

class BasePixel:
    def __init__(self, x:int, y:int):
        self._x = x
        self._y = y
    
    @property
    def x(self):
        return self._x
    
    @x.setter
    def x(self, value: int):
        self._x = value
    
    @property
    def y(self):
        return self._y
    
    @y.setter
    def y(self, value: int):
        self._y = value
    
    def __eq__(self, other):
        if isinstance(other, BasePixel):
            return self.x == other.x and self.y == other.y
        return False

class Pixel(BasePixel):
    def __init__(self, x:int, y:int, country:Country = None, adjacent_pixels:list = None, cooldown:int = 0, can_build_boat:bool = True):
        super().__init__(x, y)
        self._country = country
        # self._country_abbreviation = country_abbreviation
        # self._color = color
        self._adjacent_pixels: list[Pixel] = [] if adjacent_pixels is None else adjacent_pixels
        self._cooldown_cap = cooldown
        self._cooldown = cooldown
        self._can_build_boat = can_build_boat
    
    @property
    def country(self):
        return self._country
    
    @country.setter
    def country(self, value:Country):
        self._country = value

    @property
    def cooldown(self):
        return self._cooldown
    
    @cooldown.setter
    def cooldown(self, value:int):
        self._cooldown = value

    @property
    def cooldown_cap(self):
        return self._cooldown_cap
    
    @cooldown_cap.setter
    def cooldown_cap(self, value:int):
        self._cap = value

    # @property
    # def country_abbreviation(self):
    #     return self._country_abbreviation
    
    # @country_abbreviation.setter
    # def country_abbreviation(self, value: str):
    #     self._country_abbreviation = value
    
    # @property
    # def color(self):
    #     return self._color
    
    # @color.setter
    # def color(self, value: tuple[int, int, int]):
    #     self._color = value
    
    @property
    def adjacent_pixels(self):
        return self._adjacent_pixels
    
    @adjacent_pixels.setter
    def adjacent_pixels(self, value: list):
        self._adjacent_pixels = value
    
    @property
    def can_build_boat(self):
        return self._can_build_boat
    
    @can_build_boat.setter
    def can_build_boat(self, value:bool):
        self._can_build_boat = value

    @staticmethod
    def copyPixel(pixel):
        return Pixel(pixel.x, pixel.y, pixel.country, [], pixel.cooldown, pixel.can_build_boat)
    
    def updateWith(self, pixel):
        self.x = pixel.x
        self.y = pixel.y
        self.country = pixel.country
        self.cooldown_cap = pixel.cooldown_cap
        self.cooldown = pixel.cooldown
        self.can_build_boat = pixel.can_build_boat
    
    def add_adjacent_pixel(self, pixel):
        self.adjacent_pixels.append(pixel)
    
    def remove_adjacent_pixel(self, pixel):
        try:
            self.adjacent_pixels.remove(pixel)
        except:
            print("Adjacent pixel not found!")
    
    def __str__(self):
        return f"[\n\tlocation: ({self.x}, {self.y})\n\tcountry: {self.country}\n\tadjacent cells: {self.adjacent_pixels}\n]"